apiserver/common/modeluser.go

```diff
@@ -13,25 +13,36 @@ import (
 	coremodel "github.com/juju/juju/core/model"
 	"github.com/juju/juju/core/permission"
 	"github.com/juju/juju/core/user"
-	"github.com/juju/juju/domain/access"
 	"github.com/juju/juju/rpc/params"
 )
 
-type accessService interface {
-	// GetModelUsers will retrieve basic information about all users with
+type modelService interface {
+	// GetModelUsers will retrieve basic information about users with
 	// permissions on the given model UUID.
-	// If the model cannot be found it will return modelerrors.NotFound.
-	// If no permissions can be found on the model it will return
-	// accesserrors.PermissionNotValid.
-	GetModelUsers(ctx context.Context, apiUser user.Name, modelUUID coremodel.UUID) ([]access.ModelUserInfo, error)
+	// If the model cannot be found it will return [modelerrors.NotFound].
+	GetModelUsers(ctx context.Context, modelUUID coremodel.UUID) ([]coremodel.ModelUserInfo, error)
+	// GetModelUser will retrieve basic information about the specified model
+	// user.
+	// If the model cannot be found it will return [modelerrors.NotFound].
+	// If the user cannot be found it will return
+	// [modelerrors.UserNotFoundOnModel].
```

Comment:
That makes sense, the interface should only say what it needs rather than what the original function does.

I agree with you very much, it feels wrong that the error interface not at all enforced. I've come accross a lot of places where we check for the wrong error (e.g. an `errors.NotFound` rather than a `modelerrors.NotFound`). I would be good to look around and see if anyone else has come up with a solution for this.

Reply:
I wonder if we could do something like change the return types of the interface. Rather than returning an error, we return a `GetModelUsersError` which can only be one of the errors returned by `GetModelUsers`. Or something like that

Reply:
Yeah, returning something else than `error` for an error look like a golang antipattern. AMHO the solution is more somewhere around wrapping error practices.